### PR TITLE
More data array and dataset reduction operations

### LINKED ISSUE #2575: Add `DataArray.min()` and `max()`
Also see #2224, to avoid duplicating too much code.

### PR DESCRIPTION
**Pull request by jl-wynen**
Fixes #2575

Also adds `nanmax`, `nanmin`, `all`, `any`. And Corresponding functions for `Dataset`.

Relies on #2591

### CODE REVIEW

File: tests/core/reduction_test.py [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+def test_sum_dataset_with_coords():
+    d = sc.Dataset(data={
+        'a':
+        sc.Variable(dims=['x', 'y'], values=np.arange(6, dtype=np.int64).reshape(2, 3)),
+        'b':
+        sc.Variable(dims=['y'], values=np.arange(3, dtype=np.int64))
+    },
+                   coords={
+                       'x':
+                       sc.Variable(dims=['x'], values=np.arange(2, dtype=np.int64)),
```
Comment (SimonHeybrock) [inline tests/core/reduction_test.py:51]: Could use `sc.arange`.

File: lib/dataset/test/logical_reduction_test.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  a.masks().set("mask", mask);
+
+  const auto all_x = makeVariable<bool>(Dimensions{Dim::Y, 3}, units::m,
+                                        Values{true, true, false});
+  const auto all_y =
+      makeVariable<bool>(Dimensions{Dim::X, 2}, units::m, Values{false, false});
+  EXPECT_EQ(all(a, Dim::X).data(), all_x);
+  EXPECT_EQ(all(a, Dim::Y).data(), all_y);
+  EXPECT_FALSE(all(a, Dim::X).masks().contains("mask"));
+  EXPECT_TRUE(all(a, Dim::Y).masks().contains("mask"));
```
Comment (SimonHeybrock) [inline lib/dataset/test/logical_reduction_test.cpp:30]: This is a good example where we would now prefer writing clearly named tests, indicating what we are testing, something like `mask_along_reduction_dim_is_dropped` and `mask_along_preserved_dim_is_preserved`.

File: lib/dataset/variable_reduction.cpp [hunk 3]
```diff
@@ -9,27 +9,68 @@
...
 
 namespace scipp::dataset {
 
-Variable sum(const Variable &var, const Dim dim, const Masks &masks) {
+namespace {
+// Uses variable::special like but constructs only a scalar
+// instead of a full-sized array.
+Variable mask_fill(const Variable &prototype, const FillValue fill) {
+  return special_like(zero_like(prototype), fill);
+}
```
Comment (SimonHeybrock) [inline lib/dataset/variable_reduction.cpp:23]: Name is a bit confusing. `special_scalar_like`, maybe?

**Review by SimonHeybrock — COMMENTED** [review]
Looks good overall, minor comments below.

File: lib/dataset/variable_reduction.cpp (hunk 3 above)
Comment (SimonHeybrock) [reply lib/dataset/variable_reduction.cpp:23]: Actually it is used only once, just remove it?

File: lib/cmake/scipp-functions.cmake [hunk 4]
```diff
@@ -86,3 +86,15 @@ scipp_function("unary" bins bin_sizes SKIP_VARIABLE BASE_INCLUDE dataset/bins.h)
...
+scipp_function("reduction" reduction sum SKIP_VARIABLE)
+scipp_function("reduction" reduction nansum SKIP_VARIABLE)
+scipp_function("reduction" reduction max SKIP_VARIABLE)
+scipp_function("reduction" reduction nanmax SKIP_VARIABLE)
+scipp_function("reduction" reduction min SKIP_VARIABLE)
+scipp_function("reduction" reduction nanmin SKIP_VARIABLE)
+scipp_function("reduction" reduction all SKIP_VARIABLE)
+scipp_function("reduction" reduction any SKIP_VARIABLE)
+scipp_function("reduction" reduction mean SKIP_VARIABLE SKIP_DATASET)
+scipp_function("reduction" reduction nanmean SKIP_VARIABLE SKIP_DATASET)
```
[the remark this answers is not in the document]
Comment (jl-wynen) [reply lib/cmake/scipp-functions.cmake:99]: Yes

File: lib/dataset/mean.cpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+#include "scipp/dataset/mean.h"
+#include "scipp/dataset/astype.h"
+#include "scipp/dataset/math.h" // needed by operations_common.h
+#include "scipp/dataset/special_values.h"
+#include "scipp/dataset/sum.h"
+
+#include "../variable/operations_common.h"
+#include "dataset_operations_common.h"
+
+using scipp::variable::reduce_all_dims;
```
[the remark this answers is not in the document]
Comment (jl-wynen) [reply lib/dataset/mean.cpp:14]: removed

File: lib/dataset/nanmean.cpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+#include "scipp/dataset/astype.h"
+#include "scipp/dataset/math.h" // needed by operations_common.h
+#include "scipp/dataset/nansum.h"
+#include "scipp/dataset/special_values.h"
+#include "scipp/dataset/sum.h"
+
+#include "../variable/operations_common.h"
+#include "dataset_operations_common.h"
+
+using scipp::variable::reduce_all_dims;
```
[the remark this answers is not in the document]
Comment (jl-wynen) [reply lib/dataset/nanmean.cpp:15]: removed

File: lib/dataset/variable_reduction.cpp (hunk 3 above)
Comment (jl-wynen) [reply lib/dataset/variable_reduction.cpp]: Renamed it because I think that it (with the comment) helps to understand the purpose.

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**